README.md
# SCD4X C02 Sensor

[![Build Status](https://img.shields.io/github/actions/workflow/status/pimoroni/scd4x-python/test.yml?branch=main)](https://github.com/pimoroni/scd4x-python/actions/workflows/test.yml)
[![Coverage Status](https://coveralls.io/repos/github/pimoroni/scd4x-python/badge.svg?branch=master)](https://coveralls.io/github/pimoroni/scd4x-python?branch=master)
[![PyPi Package](https://img.shields.io/pypi/v/scd4x.svg)](https://pypi.python.org/pypi/scd4x)
[![Python Versions](https://img.shields.io/pypi/pyversions/scd4x.svg)](https://pypi.python.org/pypi/scd4x)

# Pre-requisites

You must enable:

* i2c: `sudo raspi-config nonint do_i2c 0`

You can optionally run `sudo raspi-config` or the graphical Raspberry Pi Configuration UI to enable interfaces.

# Installing

Stable library from PyPi:

* Just run `pip3 install scd4x`

In some cases you may need to use `sudo` or install pip with: `sudo apt install python3-pip`

Latest/development library from GitHub:

* `git clone https://github.com/pimoroni/scd4x-python`
* `cd scd4x-python`
* `sudo ./install.sh`



CHANGELOG.md
0.0.2
-----

* Fix byte packing when sending data (unsigned char instead of signed char).
* Fix condition in set_temperature_offset.
* Fix reading data in get_temperature_offset.
* Add get/set_automatic_self_calibration_enabled.
* Add persist_settings.
* Repackage to latest boilerplate.

0.0.1
-----

* Initial Release


tests/test_scd4x.py
import time
from unittest import mock

import pytest


def test_rdwr(smbus2, scd4x):
    smbus2.i2c_msg.read().__iter__.return_value = [
        0xF8,
        0x96,
        0x31,
        0x9F,
        0x07,
        0xC2,
        0x3B,
        0xBE,
        0x89,
    ]
    sensor = scd4x.SCD4X()

    # Send Command
    smbus2.i2c_msg.write.reset_mock()
    smbus2.i2c_msg.read().__iter__.return_value = []
    assert sensor.rdwr(0xCAFE) == []
    smbus2.i2c_msg.write.assert_called_once_with(0x62, b"\xca\xfe")

    # Send Command with Delay
    smbus2.i2c_msg.write.reset_mock()
    smbus2.i2c_msg.read().__iter__.return_value = []
    start = time.monotonic()
    assert sensor.rdwr(0xCAFE, delay=500) == []
    elapsed = time.monotonic() - start
    assert elapsed > 0.5
    smbus2.i2c_msg.write.assert_called_once_with(0x62, b"\xca\xfe")

    # Read One Byte
    smbus2.i2c_msg.write.reset_mock()
    smbus2.i2c_msg.read().__iter__.return_value = [0xF8, 0x96, 0x31]
    assert sensor.rdwr(0xCAFE, response_length=1) == 0xF896
    smbus2.i2c_msg.write.assert_called_once_with(0x62, b"\xca\xfe")

    # Read Three Byte
    smbus2.i2c_msg.write.reset_mock()
    smbus2.i2c_msg.read().__iter__.return_value = [
        0xF8,
        0x96,
        0x31,
        0x9F,
        0x07,
        0xC2,
        0x3B,
        0xBE,
        0x89,
    ]
    assert sensor.rdwr(0xCAFE, response_length=3) == [0xF896, 0x9F07, 0x3BBE]
    smbus2.i2c_msg.write.assert_called_once_with(0x62, b"\xca\xfe")

    # Write
    smbus2.i2c_msg.write.reset_mock()
    smbus2.i2c_msg.read().__iter__.return_value = []
    assert sensor.rdwr(0xCAFE, value=0x4711) == []
    smbus2.i2c_msg.write.assert_called_once_with(0x62, b"\xca\xfe\x47\x11\xd4")

    # Read and Write
    smbus2.i2c_msg.write.reset_mock()
    smbus2.i2c_msg.read().__iter__.return_value = [0xF8, 0x96, 0x31]
    assert sensor.rdwr(0xCAFE, value=0x4711, response_length=1) == 0xF896
    smbus2.i2c_msg.write.assert_called_once_with(0x62, b"\xca\xfe\x47\x11\xd4")

    # Read with invalid CRC
    smbus2.i2c_msg.write.reset_mock()
    smbus2.i2c_msg.read().__iter__.return_value = [0xF8, 0x96, 0x00]
    with pytest.raises(ValueError):
        sensor.rdwr(0xCAFE, response_length=1)


def test_reset(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.reset()
    sensor.rdwr.assert_called_once_with(scd4x.SOFT_RESET, delay=20)


def test_factory_reset(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.factory_reset()
    assert sensor.rdwr.call_args_list == [
        mock.call(scd4x.STOP_PERIODIC_MEASUREMENT, delay=500),
        mock.call(scd4x.FACTORY_RESET, delay=1200),
    ]


def test_self_test(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    # OK
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = 0
    sensor.self_test()
    assert sensor.rdwr.call_args_list == [
        mock.call(scd4x.STOP_PERIODIC_MEASUREMENT, delay=500),
        mock.call(scd4x.SELF_TEST, response_length=1, delay=10000),
    ]

    # Failure
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = 1
    with pytest.raises(RuntimeError):
        sensor.self_test()
    assert sensor.rdwr.call_args_list == [
        mock.call(scd4x.STOP_PERIODIC_MEASUREMENT, delay=500),
        mock.call(scd4x.SELF_TEST, response_length=1, delay=10000),
    ]


def test_measure(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    # OK
    start = time.time()
    sensor.rdwr.reset_mock(return_value=True, side_effect=True)
    sensor.rdwr.side_effect = (
        0x8000,  # No Data Ready
        0x8006,  # Data Ready
        [0x01F4, 0x6667, 0x5EB9],
    )
    co2, temperature, relative_humidity, timestamp = sensor.measure()
    end = time.time()
    assert co2 == 500
    assert round(temperature, 2) == 25.00
    assert round(relative_humidity, 2) == 37.00
    assert end > timestamp > start
    assert sensor.rdwr.call_args_list == [
        mock.call(scd4x.DATA_READY, response_length=1, delay=1),
        mock.call(scd4x.DATA_READY, response_length=1, delay=1),
        mock.call(scd4x.READ_MEASUREMENT, response_length=3, delay=1),
    ]

    # No Data, Non Blocking
    sensor.rdwr.reset_mock(return_value=True, side_effect=True)
    sensor.rdwr.side_effect = (0x8000,)  # No Data Ready
    assert sensor.measure(blocking=False) is None
    assert sensor.rdwr.call_args_list == [
        mock.call(scd4x.DATA_READY, response_length=1, delay=1),
    ]

    # No Data, Timeout
    sensor.rdwr.reset_mock(return_value=True, side_effect=True)
    sensor.rdwr.return_value = 0x8000  # No Data Ready
    start = time.monotonic()
    with pytest.raises(RuntimeError):
        sensor.measure(timeout=1)
    elapsed = time.monotonic() - start
    assert elapsed > 1
    assert (
        sensor.rdwr.call_args_list
        == [
            mock.call(scd4x.DATA_READY, response_length=1, delay=1),
        ]
        * 11
    )


def test_data_ready(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    # OK
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = 0x8006
    assert sensor.data_ready() is True
    sensor.rdwr.assert_called_once_with(scd4x.DATA_READY, response_length=1, delay=1)

    # Failure
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = 0x8000
    assert sensor.data_ready() is False
    sensor.rdwr.assert_called_once_with(scd4x.DATA_READY, response_length=1, delay=1)


def test_get_serial_number(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = [0xF896, 0x9F07, 0x3BBE]
    assert sensor.get_serial_number() == 0xF8969F073BBE
    sensor.rdwr.assert_called_once_with(scd4x.SERIAL_NUMBER, response_length=3, delay=1)


def test_start_periodic_measurement(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    # Normal
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.start_periodic_measurement()
    sensor.rdwr.assert_called_once_with(scd4x.START_PERIODIC_MEASUREMENT)

    # Low Power
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.start_periodic_measurement(low_power=True)
    sensor.rdwr.assert_called_once_with(scd4x.START_LOW_POWER_PERIODIC_MEASUREMENT)


def test_stop_periodic_measurement(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.stop_periodic_measurement()
    sensor.rdwr.assert_called_once_with(scd4x.STOP_PERIODIC_MEASUREMENT, delay=500)


def test_set_ambient_pressure(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.set_ambient_pressure(mock.sentinel.value)
    sensor.rdwr.assert_called_once_with(scd4x.SET_PRESSURE, value=mock.sentinel.value)


def test_set_temperature_offset(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.set_temperature_offset(5.4)
    sensor.rdwr.assert_called_once_with(scd4x.SET_TEMP_OFFSET, value=0x07E6)


def test_get_temperature_offset(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = 0x0912
    assert round(sensor.get_temperature_offset(), 2) == 6.2
    sensor.rdwr.assert_called_once_with(
        scd4x.GET_TEMP_OFFSET, response_length=1, delay=1
    )


def test_set_altitude(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.set_altitude(mock.sentinel.value)
    sensor.rdwr.assert_called_once_with(scd4x.SET_ALTITUDE, value=mock.sentinel.value)


def test_get_altitude(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = mock.sentinel.value
    assert sensor.get_altitude() == mock.sentinel.value
    sensor.rdwr.assert_called_once_with(scd4x.GET_ALTITUDE, response_length=1, delay=1)


def test_set_automatic_self_calibration_enabled(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    # Setting True
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.set_automatic_self_calibration_enabled(True)
    sensor.rdwr.assert_called_once_with(scd4x.SET_ASCE, value=1)

    # Setting False
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.set_automatic_self_calibration_enabled(False)
    sensor.rdwr.assert_called_once_with(scd4x.SET_ASCE, value=0)


def test_get_automatic_self_calibration_enabled(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    # Getting True
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = 1
    assert sensor.get_automatic_self_calibration_enabled() is True
    sensor.rdwr.assert_called_once_with(scd4x.GET_ASCE, response_length=1, delay=1)

    # Getting False
    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = 0
    assert sensor.get_automatic_self_calibration_enabled() is False
    sensor.rdwr.assert_called_once_with(scd4x.GET_ASCE, response_length=1, delay=1)


def test_persist_settings(smbus2, scd4x):
    scd4x.SCD4X.rdwr = mock.MagicMock()
    sensor = scd4x.SCD4X()

    sensor.rdwr.reset_mock()
    sensor.rdwr.return_value = []
    sensor.persist_settings()
    sensor.rdwr.assert_called_once_with(scd4x.PERSIST_SETTINGS, delay=800)


examples/self-test.py
#!/usr/bin/env python3
from scd4x import SCD4X

device = SCD4X(quiet=False)
device.self_test()

print("Self test successful!")


library/tests/conftest.py
import sys
from unittest import mock

import pytest


@pytest.fixture(scope="function", autouse=False)
def smbus2():
    """Mock smbus module."""
    sys.modules["smbus2"] = mock.MagicMock()
    yield sys.modules["smbus2"]
    del sys.modules["smbus2"]


@pytest.fixture(scope="function", autouse=False)
def scd4x():
    """Import scd4x module."""
    import scd4x

    yield scd4x
    del sys.modules["scd4x"]


examples/bargraph.py
#!/usr/bin/env python3
import sys

from scd4x import SCD4X

BAR_CHAR = u'\u2588'

ANSI_COLOR_RED = '\x1b[31m'
ANSI_COLOR_GREEN = '\x1b[32m'
ANSI_COLOR_YELLOW = '\x1b[33m'
ANSI_COLOR_BLUE = '\x1b[34m'
ANSI_COLOR_MAGENTA = '\x1b[35m'
ANSI_COLOR_BLACK = '\x1b[30m'
ANSI_COLOR_RESET = '\x1b[0m'


colours = [ANSI_COLOR_BLUE, ANSI_COLOR_GREEN, ANSI_COLOR_YELLOW, ANSI_COLOR_RED, ANSI_COLOR_MAGENTA]

BAR_WIDTH = 80

min_co2 = 1 << 32
max_co2 = 0

min_temperature = 1 << 32
max_temperature = 0

device = SCD4X()
device.start_periodic_measurement()

try:
    while True:
        co2, temperature, relative_humidity, timestamp = device.measure()

        min_co2 = min(co2 - 100, min_co2)
        max_co2 = max(co2 + 100, max_co2)
        min_temperature = min(temperature - 5, min_temperature)
        max_temperature = max(temperature + 5, max_temperature)

        if max_co2 - min_co2 == 0 or max_temperature - min_temperature == 0:
            continue

        t_scale = min((temperature - min_temperature) / (max_temperature - min_temperature), 1.0)
        p_scale = min((co2 - min_co2) / (max_co2 - min_co2), 1.0)
        t_colour = colours[int((len(colours) - 1) * t_scale)]
        p_colour = colours[int((len(colours) - 1) * p_scale)]
        t_bar = BAR_CHAR * int(BAR_WIDTH * t_scale)
        c_bar = BAR_CHAR * int(BAR_WIDTH * p_scale)

        t_bar += ANSI_COLOR_BLACK + (BAR_CHAR * (BAR_WIDTH - len(t_bar)))
        c_bar += ANSI_COLOR_BLACK + (BAR_CHAR * (BAR_WIDTH - len(c_bar)))

        t_bar = t_colour + t_bar + ANSI_COLOR_RESET
        c_bar = p_colour + c_bar + ANSI_COLOR_RESET

        t_reading = "{:.4f}c".format(temperature).ljust(BAR_WIDTH + 14)
        c_reading = "{:.0f}PPM".format(co2).ljust(BAR_WIDTH + 14)

        sys.stdout.write('\x1b[0;1H')
        sys.stdout.write(u"""{title}
{blank}
Temperature: {t_bar}
{t_reading}
CO2:         {c_bar}
{c_reading}
{blank}
""".format(
            title="SCD4X Sensor".ljust(BAR_WIDTH + 14, " "),
            t_bar=t_bar,
            c_bar=c_bar,
            t_reading=t_reading,
            c_reading=c_reading,
            blank=" " * (BAR_WIDTH + 14)
        ))
        sys.stdout.flush()

except KeyboardInterrupt:
    pass


scd4x/__init__.py
import struct
import time

from smbus2 import SMBus, i2c_msg

__version__ = "0.0.2"


SOFT_RESET = 0x3646
FACTORY_RESET = 0x3632
FORCE_RECALIBRATION = 0x362F
SELF_TEST = 0x3639
DATA_READY = 0xE4B8
STOP_PERIODIC_MEASUREMENT = 0x3F86
START_PERIODIC_MEASUREMENT = 0x21B1
START_LOW_POWER_PERIODIC_MEASUREMENT = 0x21AC
READ_MEASUREMENT = 0xEC05
SERIAL_NUMBER = 0x3682
GET_TEMP_OFFSET = 0x2318
SET_TEMP_OFFSET = 0x241D
GET_ALTITUDE = 0x2322
SET_ALTITUDE = 0x2427
SET_PRESSURE = 0xE000
PERSIST_SETTINGS = 0x3615
GET_ASCE = 0x2313
SET_ASCE = 0x2416


DEFAULT_I2C_ADDRESS = 0x62


class SCD4X:
    def __init__(self, address=None, quiet=True):
        self.co2 = 0
        self.temperature = 0
        self.relative_humidity = 0

        if address is None:
            address = DEFAULT_I2C_ADDRESS
        self.address = address

        self.bus = SMBus(1)

        self.stop_periodic_measurement()

        serial = self.get_serial_number()

        if not quiet:
            print(f"SCD4X, Serial: {serial:06x}")

    def rdwr(self, command, value=None, response_length=0, delay=0):
        if value is not None:
            msg_w = i2c_msg.write(
                self.address, struct.pack(">HHB", command, value, self.crc8(value))
            )
        else:
            msg_w = i2c_msg.write(self.address, struct.pack(">H", command))

        self.bus.i2c_rdwr(msg_w)

        time.sleep(delay / 1000.0)

        response_length *= 3

        if response_length > 0:
            msg_r = i2c_msg.read(self.address, response_length)
            self.bus.i2c_rdwr(msg_r)

            result = list(msg_r)
            data = []
            for chunk in range(0, len(result), 3):
                if self.crc8(result[chunk : chunk + 2]) != result[chunk + 2]:
                    raise ValueError("ICP10125: Invalid CRC8 in response.")
                data.append((result[chunk] << 8) | result[chunk + 1])
            if len(data) == 1:
                return data[0]
            else:
                return data

        return []

    def reset(self):
        """Resets to user settings from EEPROM"""
        self.rdwr(SOFT_RESET, delay=20)

    def factory_reset(self):
        """Reset to factory fresh condition.

        Resets user config in EEPROM.

        """
        self.stop_periodic_measurement()
        self.rdwr(FACTORY_RESET, delay=1200)

    def self_test(self):
        self.stop_periodic_measurement()
        response = self.rdwr(SELF_TEST, response_length=1, delay=10000)
        if response > 0:
            raise RuntimeError("Self test failed!")

    def measure(self, blocking=True, timeout=10):
        t_start = time.time()
        while not self.data_ready():
            if not blocking:
                return
            if time.time() - t_start > timeout:
                raise RuntimeError("Timeout waiting for data ready.")
            time.sleep(0.1)

        response = self.rdwr(READ_MEASUREMENT, response_length=3, delay=1)
        self.co2 = response[0]
        self.temperature = -45 + 175.0 * response[1] / (1 << 16)
        self.relative_humidity = 100.0 * response[2] / (1 << 16)

        return self.co2, self.temperature, self.relative_humidity, time.time()

    def data_ready(self):
        response = self.rdwr(DATA_READY, response_length=1, delay=1)
        return (response & 0x07FF) != 0

    def get_serial_number(self):
        response = self.rdwr(SERIAL_NUMBER, response_length=3, delay=1)
        return (response[0] << 32) | (response[1] << 16) | response[2]

    def start_periodic_measurement(self, low_power=False):
        if low_power:
            self.rdwr(START_LOW_POWER_PERIODIC_MEASUREMENT)
        else:
            self.rdwr(START_PERIODIC_MEASUREMENT)

    def stop_periodic_measurement(self):
        self.rdwr(STOP_PERIODIC_MEASUREMENT, delay=500)

    def set_ambient_pressure(self, ambient_pressure):
        self.rdwr(SET_PRESSURE, value=ambient_pressure)

    def set_temperature_offset(self, temperature_offset):
        if temperature_offset > 374:
            raise ValueError("Temperature offset must be <= 374c")
        offset = int(temperature_offset * (1 << 16) / 175)
        self.rdwr(SET_TEMP_OFFSET, value=offset)

    def get_temperature_offset(self):
        response = self.rdwr(GET_TEMP_OFFSET, response_length=1, delay=1)
        return 175.0 * response / (1 << 16)

    def set_altitude(self, altitude):
        self.rdwr(SET_ALTITUDE, value=altitude)

    def get_altitude(self):
        return self.rdwr(GET_ALTITUDE, response_length=1, delay=1)

    def set_automatic_self_calibration_enabled(self, value):
        self.rdwr(SET_ASCE, value=int(value))

    def get_automatic_self_calibration_enabled(self):
        return bool(self.rdwr(GET_ASCE, response_length=1, delay=1))

    def persist_settings(self):
        self.rdwr(PERSIST_SETTINGS, delay=800)

    def crc8(self, data, polynomial=0x31):
        if isinstance(data, int):
            data = [(data >> 8) & 0xFF, data & 0xFF]
        result = 0xFF
        for byte in data:
            result ^= byte
            for bit in range(8):
                if result & 0x80:
                    result <<= 1
                    result ^= polynomial
                else:
                    result <<= 1
        return result & 0xFF
